### PR TITLE
Fix DHCP Option set

### PR DESCRIPTION
**Pull request by JackyCZJ**

### CODE REVIEW

**Comment by hzhou8** [comment]
@JackyCZJ Thanks for the patch! It would be helpful if you could add commit message to describe briefly what's the problem and how is it solved.

**Comment by JackyCZJ** [comment]
Like :" Fixd: Use uuid instead of cidr as index to set DHCP Option . "?  my poor English lol

**Comment by hzhou8** [comment]
@JackyCZJ it is better. But in general, if the title doesn't provide enough details, we should add more information in the commit message body. If you are fixing something, briefly describe the problem so that reviewer quickly knows the purpose. Briefly describe the solution would also be helpful if the solution to the problem is not obvious.

Here is an example (maybe not perfect, but you can see what I meant):

commit aa10a9506bb5e4abf03508c2ffabd09b05868024
Author: Han Zhou <hzhou8@ebay.com>
Date:   Fri Mar 8 11:52:01 2019 -0800

    Fix test init.
    
    OVS sandbox NB DB socket name changed from ovn-nb.sock to nb1.ovsdb,
    which causes existing test initialization failing. This patch makes
    it smarter to read from $OVN_NB_DB env.

**Comment by hzhou8** [comment]
@vtolstov added the DHCP APIs initially. Need him to confirm why not using uuid, or any concern about using uuid.

File: dhcp_options.go [hunk 1]
```diff
@@ -91,26 +91,43 @@ func (odbi *ovndb) dhcpOptionsAddImp(cidr string, options map[string]string, ext
...
+		if err != nil {
+			return nil, err
+		}
+		row["external_ids"] = oMap
+	}
 	condition := libovsdb.NewCondition("_uuid", "==", libovsdb.UUID{dhcpUUID})
 
 	mutateOp := libovsdb.Operation{
-		Op:        opMutate,
+		Op:        opUpdate,
```
Comment (hzhou8) [inline dhcp_options.go]: The original purpose of this API (I think) is to add option key-value pairs for existing cidr (and external_ids). This patch is changing it to a complete replacement of all options and external_ids. Could you explain the motivation first?

Comment (vtolstov) [reply dhcp_options.go]: if we follow ovn-nbctl commands, we don't eed to replace, but append absent, and change exiting options.

**Comment by vtolstov** [comment]
my mistake that initially i think that cidr is uniq, but later i saw that i can have many options created in ovnnb and assign needed options to port via uuid. So we don't need to refer options via cidr, but via uniq uuid of such options.

File: dhcp_options.go [hunk 2]
```diff
@@ -91,26 +91,43 @@ func (odbi *ovndb) dhcpOptionsAddImp(cidr string, options map[string]string, ext
...
 
-func (odbi *ovndb) dhcpOptionsSetImp(cidr string, options map[string]string, external_ids map[string]string) (*OvnCommand, error) {
-
-	row, err := newDHCPRow(cidr, nil, external_ids)
-	if err != nil {
-		return nil, err
+func (odbi *ovndb) dhcpOptionsSetImp(uuid string, options map[string]string, external_ids map[string]string) (*OvnCommand, error) {
+	row := make(OVNRow)
+	dhcpUUID := ""
+	dhcpuuidlist,err := odbi.DHCPOptionsList()
```
Comment (vtolstov) [inline dhcp_options.go]: why? you need to check only that this uuid exists, so don't over allocate slice for all options.
you already have uuid, so check for ok in odbi.cache[tableDHCPOptions][UUID]

File: dhcp_options.go [hunk 3]
```diff
@@ -91,26 +91,43 @@ func (odbi *ovndb) dhcpOptionsAddImp(cidr string, options map[string]string, ext
...
+	if err != nil{
+		return nil, ErrorNotFound
 	}
-
-	dhcpUUID := odbi.getRowUUID(tableDHCPOptions, row)
-	if len(dhcpUUID) == 0 {
+	for _ ,v := range dhcpuuidlist{
```
Comment (vtolstov) [inline dhcp_options.go]: this is not needed

File: dhcp_options.go [hunk 4]
```diff
@@ -91,26 +91,43 @@ func (odbi *ovndb) dhcpOptionsAddImp(cidr string, options map[string]string, ext
...
+		if v.UUID == uuid{
+			dhcpUUID = uuid
+			break
+		}
+	}
+	if dhcpUUID == ""{
 		return nil, ErrorNotFound
 	}
+	if options != nil {
```
Comment (vtolstov) [inline dhcp_options.go]: may be if options is nil - return error?

File: dhcp_options.go (hunk 1 above)
Comment (JackyCZJ) [reply dhcp_options.go]: got it , wip

Comment (JackyCZJ) [reply dhcp_options.go]:
```
root@esixCell:/home/jacky# ovn-nbctl dhcp-options-set-options ac5a44a0-8c38-4f3b-a861-1e1780d6287a "server_id"="192.168.1.1" "server_mac"="54:54:54:54:54:54" "lease_time"="6000"
root@esixCell:/home/jacky# ovn-nbctl list  DHCP_Options
_uuid               : ac5a44a0-8c38-4f3b-a861-1e1780d6287a
cidr                : "192.168.0.0/24"
external_ids        : {}
options             : {lease_time="6000", server_id="192.168.1.1", server_mac="54:54:54:54:54:54"}
root@esixCell:/home/jacky# ovn-nbctl dhcp-options-set-options ac5a44a0-8c38-4f3b-a861-1e1780d6287a "server_id"="192.168.1.1"  "lease_time"="5000"
root@esixCell:/home/jacky# ovn-nbctl list  DHCP_Options
_uuid               : ac5a44a0-8c38-4f3b-a861-1e1780d6287a
cidr                : "192.168.0.0/24"
external_ids        : {}
options             : {lease_time="5000", server_id="192.168.1.1"}
```
this is ovn-nbctl  dhcp-options-set-options , it just completely replaced . To replace like key-value , we need to contrast change to update, becase options all in one field.

File: dhcp_options.go (hunk 4 above)
Comment (JackyCZJ) [reply dhcp_options.go]: maybe we can make options to "" to clean all options.....

**Comment by JackyCZJ** [comment]
> @JackyCZJ it is better. But in general, if the title doesn't provide enough details, we should add more information in the commit message body. If you are fixing something, briefly describe the problem so that reviewer quickly knows the purpose. Briefly describe the solution would also be helpful if the solution to the problem is not obvious.
> 
> Here is an example (maybe not perfect, but you can see what I meant):
> 
> commit [aa10a95](https://github.com/eBay/go-ovn/commit/aa10a9506bb5e4abf03508c2ffabd09b05868024)
> Author: Han Zhou [hzhou8@ebay.com](mailto:hzhou8@ebay.com)
> Date: Fri Mar 8 11:52:01 2019 -0800
> 
> ```
> Fix test init.
> 
> OVS sandbox NB DB socket name changed from ovn-nb.sock to nb1.ovsdb,
> which causes existing test initialization failing. This patch makes
> it smarter to read from $OVN_NB_DB env.
> ```

thanks for suggestion !

File: dhcp_options.go (hunk 2 above)
Comment (JackyCZJ) [reply dhcp_options.go]: done

File: dhcp_options.go (hunk 4 above)
Comment (JackyCZJ) [reply dhcp_options.go]: done

File: dhcp_options.go (hunk 3 above)
Comment (JackyCZJ) [reply dhcp_options.go]: donw

File: dhcp_options.go (hunk 1 above)
Comment (vtolstov) [reply dhcp_options.go]: Ok if ovn-nbctl replaces options do the same

Comment (JackyCZJ) [reply dhcp_options.go]: wait , i commit my test env .... I should fix it

**Comment by JackyCZJ** [comment]
@hzhou8 @vtolstov PTAL

File: client.go [hunk 5]
```diff
@@ -388,20 +390,24 @@ func (c *ovndb) DHCPOptionsAdd(cidr string, options map[string]string, external_
...
+	return c.dhcpOptionsGetImp(uuid)
+}
+
 func (c *ovndb) DHCPOptionsList() ([]*DHCPOptions, error) {
 	return c.dhcpOptionsListImp()
 }
 
 func (c *ovndb) LRNATAdd(lr string, ntype string, externalIp string, logicalIp string, external_ids map[string]string, logicalPortAndExternalMac ...string) (*OvnCommand, error) {
-	return c.lrNatAddImp(lr, ntype, externalIp, logicalIp,external_ids, logicalPortAndExternalMac...)
+	return c.lrNatAddImp(lr, ntype, externalIp, logicalIp, external_ids, logicalPortAndExternalMac...)
```
Comment (vtolstov) [inline client.go]: unrelated to this pr change, please use goimports with you code editor.

File: dhcp_options.go [hunk 6]
```diff
@@ -91,27 +91,37 @@ func (odbi *ovndb) dhcpOptionsAddImp(cidr string, options map[string]string, ext
...
+func (odbi *ovndb) dhcpOptionsSetImp(uuid string, options map[string]string, external_ids map[string]string) (*OvnCommand, error) {
+	row := make(OVNRow)
 
-	row, err := newDHCPRow(cidr, nil, external_ids)
+	_, ok := odbi.cache[tableDHCPOptions][uuid]
+	if !ok {
+		return nil, ErrorNotFound
+	}
+	if options == nil {
+		return nil, ErrorSchema
```
Comment (vtolstov) [inline dhcp_options.go]: use different error type for invalid args - ErrorOption

**Comment by vtolstov** [comment]
i'm fix minor suggestions and push

**Pull request merged**